File: week7/partition_equal_subset_sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int tot=0;
        for(int i=0;i<nums.size();i++) tot+=nums[i];
        if(tot%2!=0) return false;
        int x = tot/2;
        vector<bool> dp(x+1,false);
        dp[0]=true;
        for(int i=0;i<nums.size();i++){
            for(int j=x;j>=0;j--){
                if(j-nums[i]>=0 && dp[j-nums[i]]) dp[j]=true;
            }
        }
        return dp[x];
    }
};

File: Chavda_23B0368/Week_7/Partition_equal_subset_sum.java
package Week_7;

public class Partition_equal_subset_sum {
    Boolean[][] dp;
    public boolean canPartition(int[] nums) {
        int total = 0;
        for(int num : nums) total += num;

        if(total % 2 != 0)  return false;
        dp = new Boolean[nums.length][(total/2) + 1];
        return dfs(nums, 0, 0, total/2);
    }

    private boolean dfs(int[] nums, int idx, int sum, int target) {
        if(sum == target)
            return true;
        if(idx == nums.length || sum > target)  
            return false;
        if(dp[idx][sum] != null)    
            return dp[idx][sum];

        boolean take = dfs(nums, idx+1, sum+nums[idx], target);
        boolean not_take = dfs(nums, idx+1, sum, target);

        return dp[idx][sum] = take || not_take;
    }
}